returnn/tf/util/data.py

```diff
@@ -1088,25 +1099,42 @@ def get_default_new_axis_for_dim_tag(self, dim_tag):
     else:
       return self.batch_ndim  # add it at the end
 
-  def copy_add_dim_by_tag(self, dim_tag, unbroadcast=False, axis=None):
+  def copy_add_dim_by_tag(self, dim_tag, unbroadcast=False, axis=None, dyn_size=None):
     """
     :param DimensionTag dim_tag:
     :param bool unbroadcast:
     :param int|None axis:
+    :param tf.Tensor|int|None dyn_size:
+      For unbroadcast, if we do not want to rely on dim_tag.dimension.
```

Comment:
In what cases can that happen? If you know the `dyn_size`, maybe you would (should) just add it to the `dim_tag`?

Reply:
max(dim_tag.dyn_size) == dyn_size here, or not?

Reply:
That was what I was wondering about. If that is the case, we can do that.

Reply:
I think that should always be the case.

Reply:
Thanks! One issue maybe: What if it is the batch dim? Then, `dim_tag.dyn_size` and `dim_tag.dimension` are both unset.
Or maybe should we could use `tf.shape(dim_tag.src_data.placeholder)[dim_tag.src_axis]` if that exists?
But what if the `placeholder` does not exist?

I really don't know enough about the attributes of `DimensionTag` to tell there is any other way to consistently find the dim of the batch axis, or whether we actually need to add a parameter in this case.

Reply:
If there is no good easy way to get the batch dim, we maybe should add such attribute.

In some cases, we use `LayerBase.get_recent_layer().get_batch_dim()`. But this is somewhat ugly. Although maybe a good simple workaround for now, and easy to replace later.

Reply:
Okay, thanks, sure!

One question that remains is how we should handle the (now somewhat obsolete) `data_dyn_shape` parameter of `Data.copy_compatible_to`: Just as with `Data.copy_add_dim_by_tag`, this information can be inferred using the DimensionTags of the input.
All tests also seem to pass with just entirely ignoring the `data_dyn_shape` param.
Maybe we should just deprecate and ignore it then (but keep it for backwards compatibility to the API)?

The alternative would be a somewhat ugly hack to set the `dyn_size` (or `dimension`, if its static) on the dimension tags according to `data_dyn_shape` within `Data.copy_compatible_to`.

Reply:
I assume it was not always the case that it could be inferred completely.

We should clean that up and remove the param. I don't think this API is really used externally, and internally we can simply clean that up.

Reply:
Yes, okay, I will remove it entirely (was not used much anyway). I will also remove the `out_shape` argument from `Data.get_common_shape`.